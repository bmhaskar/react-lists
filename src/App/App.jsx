import React from 'react'
import { Column } from '../components/Column';
import styled from "styled-components";
import {DragDropContext,Droppable} from 'react-beautiful-dnd'


const Container = styled.div`
  display: flex;
`;

class InnerColumns extends React.PureComponent {
    render() {
      const { index, column, taskMap } = this.props;
      const tasks = column.taskIds.map(taskId => taskMap[taskId]);
      return <Column index={index} column={column} tasks={tasks} />;
    }
  }
class App extends React.Component {
    static defaultProps = {initialData: {}}
    state = this.props.intialData ;
    onDragEnd = result => {
      const { destination, source, draggableId, type } = result;
      if (!destination) {
        return;
      }
      if (
        destination.droppableId === source.droppableId &&
        destination.index === source.index
      ) {
        return;
      }
  
      if (type === "column") {
        const newColumnOrder = Array.from(this.state.columnOrder);
        newColumnOrder.splice(source.index, 1);
        newColumnOrder.splice(destination.index, 0, draggableId);
        const newState = {
          ...this.state,
          columnOrder: newColumnOrder
        };
        this.setState(newState);
        return;
      }
      const start = this.state.columns[source.droppableId];
      const finish = this.state.columns[destination.droppableId];
      if (start === finish) {
        const newTaskIds = Array.from(start.taskIds);
        newTaskIds.splice(source.index, 1);
        newTaskIds.splice(destination.index, 0, draggableId);
        const newColumn = {
          ...start,
          taskIds: newTaskIds
        };
        const newState = {
          ...this.state,
          columns: {
            ...this.state.columns,
            [newColumn.id]: newColumn
          }
        };
  
        this.setState(newState);
        return;
      }
  
      const newStartTaskIds = Array.from(start.taskIds);
      newStartTaskIds.splice(source.index, 1);
      const newStartColumn = {
        ...start,
        taskIds: newStartTaskIds
      };
      const newFinishTaskIds = Array.from(finish.taskIds);
      newFinishTaskIds.splice(destination.index, 0, draggableId);
      const newFinishColumn = {
        ...finish,
        taskIds: newFinishTaskIds
      };
  
      const newState = {
        ...this.state,
        columns: {
          ...this.state.columns,
          [newStartColumn.id]: newStartColumn,
          [newFinishColumn.id]: newFinishColumn
        }
      };
  
      this.setState(newState);
    };
    render() {
      return (
        <DragDropContext onDragEnd={this.onDragEnd}>
          <Droppable
            droppableId="all-columns"
            direction="horizontal"
            type="column"
          >
            {provided => (
              <Container ref={provided.innerRef} {...provided.droppableProps}>
                {this.state.columnOrder.map((columnId, index) => {
                  const column = this.state.columns[columnId];
                  return (
                    <InnerColumns
                      key={columnId}
                      column={column}
                      taskMap={this.state.tasks}
                      index={index}
                    />
                  );
                })}
                {provided.placeholder}
              </Container>
            )}
          </Droppable>
        </DragDropContext>
      );
    }
  }
  export default App;