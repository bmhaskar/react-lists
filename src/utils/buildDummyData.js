import intialData from "../intial-data";
const buildDummyData = function(intialData) {
  const getFirstColumn = () => {
    return intialData.columns[intialData.columnOrder[0]];
  };
  const getLastColumn = () => {
    return intialData.columns[
      intialData.columnOrder[intialData.columnOrder.length - 1]
    ];
  };
  const getColumnTitles = () =>
    intialData.columnOrder.map(columnId => intialData.columns[columnId].title);
  const toJSON = () => intialData;
  return {
    toJSON,
    getFirstColumn,
    getColumnTtitles: getColumnTitles,
    getLastColumn
  };
};
export const testData = new buildDummyData(intialData);
