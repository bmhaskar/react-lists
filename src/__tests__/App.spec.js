import React from "react";
import { render, cleanup } from "react-testing-library";
import "jest-dom/extend-expect";
import App from "../App";

import { testData } from "../utils/buildDummyData";

afterEach(cleanup);
test("Test if app renders fine with dummy data", () => {
  const { getAllByTestId } = render(<App intialData={testData.toJSON()} />);
  const titles = getAllByTestId("column-title").map(
    titleNode => titleNode.textContent
  );
  expect(titles).toEqual(testData.getColumnTtitles());
});

test("Test if app renders fine without data", () => {
  const dom = render(<App />);
  expect(dom.container).toBeInTheDocument();
  expect(dom.asFragment()).toMatchSnapshot();
});
