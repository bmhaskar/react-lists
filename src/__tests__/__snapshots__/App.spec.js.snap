// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test if app renders fine without data 1`] = `
<DocumentFragment>
  <h1
    class="sc-EHOje dybXsK"
  >
    React lists
  </h1>
  <div
    class="sc-ifAKCX bUJsJm"
    data-react-beautiful-dnd-droppable="1"
  />
</DocumentFragment>
`;
